### PR TITLE
Generate pyodide `repodata.json`

### PR DESCRIPTION
**Pull request by bollwyvl**
## References

- includes jupyterlite/jupyterlite#978
- as suggested in https://github.com/jupyterlite/pyodide-kernel/issues/14
- jupyterlite/jupyterlite#904
- jupyterlite/pyodide-kernel#60
- jupyterlite/jupyterlite#965 (as long as the `import` is in the same cell as the magic)

## Code changes

- [x] add pyodide-compatible `repodata.json` generation
  - [x] extract data from `top_level.txt` (`setuptools` wheels)
  - [x] detect from `RECORD` by
    - [x] `site-packages/<importable>.py`
    - [x] shortest unique `__init__.py` paths (curse you namespace packages)
  - > can be seen [on RTD](https://jupyterlite--979.org.readthedocs.build/en/979/_static/pypi/repodata.json) 
- [x] include repodata in build
- [x] use repodata to patch pyodide auto import functions
- [x] add settings to enable skipping/patching install of packages e.g. `jedi`, `parso` (saves 2mb before gzip)
- [x] mock `jupyterlab_widgets` (saves another 300kb before gzip) 
- [x] mock `prompt_toolkit` (saves another 1.1mb before gzip) 
- [x] remove most `%piplite` calls from examples 
  - [x] except for the one _about_ piplite for arbitrary wheels from the internet
- [x] hide all of the above behind an opt-in flag for now, 
  - `--piplite-install-on-import` 
  - `PyodideAddon.install_on_import`
- [x] ~~add `repodata_patches` to pyolite plugin config~~
  - > i couldn't make this work properly
- [x] tests
  - [x] use `127.0.0.1` everywhere instead of `localhost`, as some CI machines have started doing more ipv6 stuff that some of the tools (`karma`) don't know how to work with by default
- [x] docs

## User-facing changes

- viewers of JupyterLite sites with pre-configured metadata 
  - will not see non-portable magics in `pyolite` examples
  - kernels should start somewhat faster due to
    - making fewer API requests
    - loading shims for `jedi` and `prompt_toolkit` which pretty much don't work in the browser anyway)
- site builders will be able to 
  - replace `pyodide`-shipped modules (the demo currently shows this, shipping a newer `pygments`
    - but _might_ not work automagically for exotic wheels

## Backwards-incompatible changes

- I'm still not entirely sure what is different between `micropip.install` and `pyodide_js.loadPackage`
  - hence this behavior is currently opt-in

## Notes

- `repodata.json` can't handle multiple versions of the same package
- finding exported names is trash, and maybe a fool's errand for the general case

### CODE REVIEW

**Comment by bollwyvl** [comment]
the folium example is broken, as we're not finding `charset-normalizer` for some reason.

**Comment by bollwyvl** [comment]
Some crude performance on: 
- opening `lab/index.html` on RTD
- wait for the launcher
- clear the network cache
- open a notebook

| task| #978 + #979 | main |
|-|-|-|
| start kernel | image | image
| load `ipywidgets`, draw slider | image | image |

<10mb is :yum: .

**Comment by bollwyvl** [comment]
With `prompt_toolkit` and `jedi` out, the remaining heavy hitters:
image

Some ideas:
- 500kb: somehow patch out the need for `sqlite3`
  - we already have sqlite3-backed history disabled
    - but it's already unvendored from upstream and requires an explicit install
      - removing it would probably breaks other packages, too
- 200kb: convince `ipython` to stop shipping tests in the wheel
  - by naively deleting everything that looks like a test, this would save
  - as it no longer ships the `iptest` command, there's really no user-serviceable reason to keep them
- 20kb: convince `traitlets` to stop shipping tests in the wheel

With those we would get under 9mb, but below that would start to see diminishing marginal returns vs complexity.

**Comment by bollwyvl** [comment]
aside from the UI test fails (which I'll look into, but likely on the weekend) I don't anticipate any additional changes, so the invite for review stands @jtpio.

**Comment by bollwyvl** [comment]
Hm, UI test fails look real, I pine for #925.

**Comment by bollwyvl** [comment]
ah-ha: absolutely can reproduce the issue. I moved the patches to before loading `micropip` and `ipykernel`, and the core `/app/lab/index.html` _doesn't_ ship the dependencies of `ipykernel`, so it fails (initially) on `wcwidth`. 

I had naively assumed this would be able to fall back to fetching with `micropip` if not available, but clearly this is not the case! Will move back to draft and think more.

**Comment by bollwyvl** [comment]
Ok, back to ready for review!

**Comment by bollwyvl** [comment]
@jtpio guess I'd like to show this at the next community call... the PR, as it stands, would be fine to demo, but would be lovely to get this in so we can do some more downstream work with it without building from source.

**Comment by jtpio** [comment]
Thanks @bollwyvl.

Yeah I was hoping to get back to https://github.com/jupyterlite/jupyterlite/pull/854 and https://github.com/jupyterlite/pyodide-kernel/pull/2 before looking at this PR since it brings quite a lot of changes and could be tedious to port over :sweat_smile:

**Comment by bollwyvl** [comment]
Welp, either way. If we're ready to start moving stuff over, that's fine, I can do the rework: really no complaints from me, as this of course would _improve_ the downstream situation.

File: app/repodata.schema.v0.json [hunk 1]
```diff
@@ -0,0 +1,107 @@
+{
+  "$schema": "http://json-schema.org/draft-07/schema",
```
Comment (jtpio) [inline app/repodata.schema.v0.json:2]: Wondering how this file will be handled once the Pyodide kernel is moved to a separate repo? (https://github.com/jupyterlite/jupyterlite/pull/854)

Comment (bollwyvl) [reply app/repodata.schema.v0.json:2]: It only gets used by `jupyter lite check`, and wouldn't _strictly_ need to be shipped with the app (or, now, extension).

We'd also move the changed chunks of `litePLuginSettings` out of the main schema into the dedicated `check`... I _do_ think it's worth shipping the main schema with the as-deployed app.

**Comment by bollwyvl** [comment]
Moved the parent issue to https://github.com/jupyterlite/pyodide-kernel.  

Humorously, the previous number (#914) became https://github.com/jupyterlite/pyodide-kernel/issues/14

**Comment by jtpio** [comment]
Closing as this will more likely happen in the `pyodide-kernel` repo: https://github.com/jupyterlite/pyodide-kernel.

Thanks!

**Pull request closed without merging**